Reapply our patch to upstream php 8.0.2

davidrans opened the pull request:
Went through the process in the README to re-apply our patch against the
newest stable release of PHP.

CC @danielbeardsley, @scbarber

danielbeardsley committed:
README: update for our build

danielbeardsley committed:
PHP Templates: add optional autescaping to echo/<?=

Background
==========
PHP doesn't have any mechanism to inject logic between templating
and final output. There is no way to filter or alter the content
that comes from code in templates like: <?= $someVar ?>

To use php as a robust templataing language, we must inject *some*
logic between templates and their output. We have chosen to make
all <?=, echo, and print statements subject to an optional
trip through escape_html_entitiles.

The functionality can be toggled with `ini_set('__auto_escape')`
and configured with `__auto_escape_flags` and
`__auto_escape_exempt_class` (see commit
2dae5d16436ce37856f6e00ca2a1b3009bb1f7ed for info about the class
name based auto-escaping exemption.

Methodology
===========
T_ECHO (echo, <?=), T_PRINT (print) now both emit a
ZEND_AST_ECHO_ESCAPE node in the syntax tree.

That's compiled to a function which emits a ZEND_ECHO_ESCAPE op code.

The op code interpretation is a dupe of ZEND_ECHO except with some
if() statements that switch the underlying function from `zend_write`
to `zend_write_escape` based on the ini settings.

zend_write_escape is a new function pointer that points to
php_escape_write.

php_escape_write is a new function that passes it's string argument
through php_escape_html_entities() (with __auto_escape_flags) before
calling the underlying php_output_write.

Use
===
This functionality allows us to safely use php straight as a
templating language with no template compilation step (as many
other templating libraries have).

See the included tests for more usage information.

Exempt Class
============
It is useful to allow some utility functions and helpers to produce
html and have it passed straight through in the template (without
being double-encoded). We accomplish this by *tagging* strings
as being HTML.

    class HtmlString implements JsonSerializable {
       protected $html = '';

       public function __construct($html) {
          $this->html = $html;
       }

       public function __toString() {
          return (string)$this->html;
       }

       public function jsonSerialize() {
          return $this->html;
       }
    }

The auto-escaping system can be configured with an:
__auto_escape_exempt_class="HtmlString"

Which allows instances of `HtmlString` to pass straight through a
template without being modified (skipping the html_entities call).
Helper functions can now return html safely and consumers don't have
to care if it is HTML or not because the auto-escaping system knows
what to do.

squash with big commit

davidrans committed:
Auto-generated files: update

The previous commit updated zend_vm_def and others that are used as
inputs to these auto-generated files.

Run php Zend/zend_vm_gen.php and commit the result.

davidrans commented on Zend/zend_vm_def.h:
```diff
@@ -1683,6 +1683,49 @@ ZEND_VM_HANDLER(136, ZEND_ECHO, CONST|TMPVAR|CV, ANY)
 }
 
+// This is copied from ZEND_ECHO above with only a small part different
+ZEND_VM_HANDLER(200, ZEND_ECHO_ESCAPE, CONST|TMPVAR|CV, ANY)
```
I had to update this from 199 -> 200 cause 199 is now used. I could pick a bigger number to help avoid future conflicts?

jarstelfox replied:
What is this number used for? Is it basically an op code? 

I found this list: https://github.com/php/php-src/blob/master/Zend/zend_vm_handlers.h

which would indicated 200 is also used?

danielbeardsley replied:
Different set of numbers.

davidrans commented on Zend/zend_vm_def.h:
```diff
@@ -1683,6 +1683,49 @@ ZEND_VM_HANDLER(136, ZEND_ECHO, CONST|TMPVAR|CV, ANY)
+ZEND_VM_HANDLER(200, ZEND_ECHO_ESCAPE, CONST|TMPVAR|CV, ANY)
+{
+	USE_OPLINE
+	zval *z;
```
Removed a variable definition here that was removed in the function this was copied from

davidrans commented on Zend/zend.c:
```diff
@@ -831,6 +835,7 @@ void zend_startup(zend_utility_functions *utility_functions) /* {{{ */
 	zend_error_cb = utility_functions->error_function;
 	zend_printf = utility_functions->printf_function;
 	zend_write = utility_functions->write_function;
+	zend_write_escape = utility_functions->write_escape_function;
```
Removed a cast here that was removed from the line above

danielbeardsley replied:
Nice!

djmetzle commented:
CR :gear:

danielbeardsley commented:
CR :+1: 

Well done! How was following the instructions? Anything we should update or add more detail to?

Also, does it compile?

davidrans commented:
> How was following the instructions?

Great! 

>  Also, does it compile?

Compiling now... done!
Testing now...

davidrans commented:
Here's the test output:
<details>
=====================================================================
TEST RESULT SUMMARY
---------------------------------------------------------------------
Exts skipped    :   46
Exts tested     :   26
---------------------------------------------------------------------

Number of tests : 16103             10960
Tests skipped   : 5143 ( 31.9%) --------
Tests warned    :    1 (  0.0%) (  0.0%)
Tests failed    :   25 (  0.2%) (  0.2%)
Expected fail   :   30 (  0.2%) (  0.3%)
Tests passed    : 10904 ( 67.7%) ( 99.5%)
---------------------------------------------------------------------
Time taken      :   92 seconds
=====================================================================

=====================================================================
EXPECTED FAILED TEST SUMMARY
---------------------------------------------------------------------
Inconsistencies when accessing protected members [Zend/tests/access_modifiers_008.phpt]  XFAIL REASON: Discussion: http://marc.info/?l=php-internals&m=120221184420957&w=2
Inconsistencies when accessing protected members - 2 [Zend/tests/access_modifiers_009.phpt]  XFAIL REASON: Discussion: http://marc.info/?l=php-internals&m=120221184420957&w=2
Test open_basedir configuration [tests/security/open_basedir_linkinfo.phpt]  XFAIL REASON: BUG: open_basedir cannot delete symlink to prohibited file. See also
bugs 48111 and 52176.
DateTime::add() -- fall type2 type3 [ext/date/tests/DateTime_add-fall-type2-type3.phpt]  XFAIL REASON: Various bugs exist
DateTime::add() -- fall type3 type2 [ext/date/tests/DateTime_add-fall-type3-type2.phpt]  XFAIL REASON: Various bugs exist
DateTime::add() -- fall type3 type3 [ext/date/tests/DateTime_add-fall-type3-type3.phpt]  XFAIL REASON: Various bugs exist
DateTime::add() -- spring type2 type3 [ext/date/tests/DateTime_add-spring-type2-type3.phpt]  XFAIL REASON: Various bugs exist
DateTime::add() -- spring type3 type2 [ext/date/tests/DateTime_add-spring-type3-type2.phpt]  XFAIL REASON: Various bugs exist
DateTime::add() -- spring type3 type3 [ext/date/tests/DateTime_add-spring-type3-type3.phpt]  XFAIL REASON: Various bugs exist
DateTime::sub() -- fall type2 type3 [ext/date/tests/DateTime_sub-fall-type2-type3.phpt]  XFAIL REASON: Various bugs exist
DateTime::sub() -- fall type3 type2 [ext/date/tests/DateTime_sub-fall-type3-type2.phpt]  XFAIL REASON: Various bugs exist
DateTime::sub() -- fall type3 type3 [ext/date/tests/DateTime_sub-fall-type3-type3.phpt]  XFAIL REASON: Various bugs exist
DateTime::diff() -- fall type2 type3 [ext/date/tests/DateTime_diff-fall-type2-type3.phpt]  XFAIL REASON: Various bugs exist
DateTime::diff() -- fall type3 type2 [ext/date/tests/DateTime_diff-fall-type3-type2.phpt]  XFAIL REASON: Various bugs exist
DateTime::sub() -- spring type2 type3 [ext/date/tests/DateTime_sub-spring-type2-type3.phpt]  XFAIL REASON: Various bugs exist
DateTime::sub() -- spring type3 type2 [ext/date/tests/DateTime_sub-spring-type3-type2.phpt]  XFAIL REASON: Various bugs exist
DateTime::sub() -- spring type3 type3 [ext/date/tests/DateTime_sub-spring-type3-type3.phpt]  XFAIL REASON: Various bugs exist
DateTime::diff() -- fall type3 type3 [ext/date/tests/DateTime_diff-fall-type3-type3.phpt]  XFAIL REASON: Various bugs exist
DateTime::diff() -- spring type2 type3 [ext/date/tests/DateTime_diff-spring-type2-type3.phpt]  XFAIL REASON: Various bugs exist
DateTime::diff() -- spring type3 type2 [ext/date/tests/DateTime_diff-spring-type3-type2.phpt]  XFAIL REASON: Various bugs exist
DateTime::diff() -- spring type3 type3 [ext/date/tests/DateTime_diff-spring-type3-type3.phpt]  XFAIL REASON: Various bugs exist
Bug #52480 (Incorrect difference using DateInterval) [ext/date/tests/bug52480.phpt]  XFAIL REASON: See https://bugs.php.net/bug.php?id=52480
RFC: DateTime and Daylight Saving Time Transitions (zone type 3, bd2) [ext/date/tests/rfc-datetime_and_daylight_saving_time-type3-bd2.phpt]  XFAIL REASON: Still not quite right
RFC: DateTime and Daylight Saving Time Transitions (zone type 3, fs) [ext/date/tests/rfc-datetime_and_daylight_saving_time-type3-fs.phpt]  XFAIL REASON: Still not quite right
Bug #42718 (unsafe_raw filter not applied when configured as default filter) [ext/filter/tests/bug42718.phpt]  XFAIL REASON: FILTER_UNSAFE_RAW not applied when configured as default filter, even with flags
Bug #67296 (filter_input doesn't validate variables) [ext/filter/tests/bug49184.phpt]  XFAIL REASON: See Bug #49184
Bug #67167: filter_var(null,FILTER_VALIDATE_BOOLEAN,FILTER_NULL_ON_FAILURE) returns null [ext/filter/tests/bug67167.02.phpt]  XFAIL REASON: Requires php_zval_filter to not use convert_to_string for all filters.
via [ext/pdo_sqlite/tests/common.phpt]
        SQLite PDO Common: PDOStatement::getColumnMeta [ext/pdo_sqlite/tests/pdo_022.phpt]  XFAIL REASON: This feature is not yet finalized, no test makes sense
Phar web-based phar with fatal error [ext/phar/tests/fatal_error_webphar.phpt]  XFAIL REASON: Uses no longer supported __autoload() function, new phar needed
updateTimestamp never called when session data is empty [ext/session/tests/bug71162.phpt]  XFAIL REASON: Current session module is designed to write empty session always. In addition, current session module only supports SessionHandlerInterface only from PHP 7.0.
=====================================================================

=====================================================================
FAILED TEST SUMMARY
---------------------------------------------------------------------
SCCP 002: Conditional Constant Propagation of non-escaping array elements [ext/opcache/tests/opt/sccp_002.phpt]
SCCP 003: Conditional Constant Propagation of non-escaping array elements [ext/opcache/tests/opt/sccp_003.phpt]
SCCP 004: Conditional Constant Propagation of non-escaping array elements [ext/opcache/tests/opt/sccp_004.phpt]
SCCP 005: Conditional Constant Propagation of non-escaping array elements [ext/opcache/tests/opt/sccp_005.phpt]
SCCP 006: Conditional Constant Propagation of non-escaping array elements [ext/opcache/tests/opt/sccp_006.phpt]
SCCP 007: Conditional Constant Propagation of non-escaping array elements [ext/opcache/tests/opt/sccp_007.phpt]
SCCP 008: Conditional Constant Propagation of non-escaping array elements [ext/opcache/tests/opt/sccp_008.phpt]
SCCP 009: Conditional Constant Propagation of non-escaping array elements [ext/opcache/tests/opt/sccp_009.phpt]
SCCP 010: Conditional Constant Propagation of non-escaping object properties [ext/opcache/tests/opt/sccp_010.phpt]
SCCP 011: Conditional Constant Propagation of non-escaping object properties [ext/opcache/tests/opt/sccp_011.phpt]
SCCP 012: Conditional Constant Propagation of non-escaping object properties [ext/opcache/tests/opt/sccp_012.phpt]
SCCP 031: Echo optimizations [ext/opcache/tests/opt/sccp_031.phpt]
SCCP 032: Yield from optimizations [ext/opcache/tests/opt/sccp_032.phpt]
SCCP 022: Invalid types [ext/opcache/tests/opt/sccp_022.phpt]
SCCP 024: Const replacing to op2 of INSTANCEOF [ext/opcache/tests/opt/sccp_024.phpt]
Test match jump table optimizer [ext/opcache/tests/match/002.phpt]
Bug #79912 (Phar::decompressFiles not working) [ext/phar/tests/bug79912.phpt]
Bug #69279 (Compressed ZIP Phar extractTo() creates garbage files) [ext/phar/tests/bug69279.phpt]
Bug #69279 (Compressed ZIP Phar extractTo() creates garbage files) [ext/phar/tests/bug69279a.phpt]
Test fsockopen() function : error conditions [ext/standard/tests/network/fsockopen_error.phpt]
Test opcode breakpoints [sapi/phpdbg/tests/breakpoints_004.phpt]
Stepping with exceptions must not be stuck at CATCH [sapi/phpdbg/tests/stepping_001.phpt]
Test breaks on HANDLE_EXCEPTION [sapi/phpdbg/tests/exceptions_003.phpt]
Test next command on function boundaries [sapi/phpdbg/tests/next_001.phpt]
file upload greater than 2G [sapi/cli/tests/upload_2G.phpt]
=====================================================================

=====================================================================
WARNED TEST SUMMARY
---------------------------------------------------------------------
Phar: bug #69958: Segfault in Phar::convertToData on invalid file [ext/phar/tests/bug69958.phpt] (warn: XFAIL section but test passes)
=====================================================================
</details>

I'm gonna build/compile/test on vanilla 8.0.2 and compare the test output...

davidrans commented:
Here's the diff:
```diff
▶  diff original patched
8c8
< Number of tests : 16100             10957
---
> Number of tests : 16103             10960
11c11
< Tests failed    :    5 (  0.0%) (  0.0%)
---
> Tests failed    :   25 (  0.2%) (  0.2%)
13c13
< Tests passed    : 10921 ( 67.8%) ( 99.7%)
---
> Tests passed    : 10904 ( 67.7%) ( 99.5%)
15c15
< Time taken      :   87 seconds       
---
> Time taken      :   92 seconds
32a33
> DateTime::sub() -- fall type3 type3 [ext/date/tests/DateTime_sub-fall-type3-type3.phpt]  XFAIL REASON: Various bugs exist
35,38d35
< DateTime::sub() -- fall type3 type3 [ext/date/tests/DateTime_sub-fall-type3-type3.phpt]  XFAIL REASON: Various bugs exist
< DateTime::diff() -- fall type3 type3 [ext/date/tests/DateTime_diff-fall-type3-type3.phpt]  XFAIL REASON: Various bugs exist
< DateTime::diff() -- spring type2 type3 [ext/date/tests/DateTime_diff-spring-type2-type3.phpt]  XFAIL REASON: Various bugs exist
< DateTime::diff() -- spring type3 type2 [ext/date/tests/DateTime_diff-spring-type3-type2.phpt]  XFAIL REASON: Various bugs exist
41a39,41
> DateTime::diff() -- fall type3 type3 [ext/date/tests/DateTime_diff-fall-type3-type3.phpt]  XFAIL REASON: Various bugs exist
> DateTime::diff() -- spring type2 type3 [ext/date/tests/DateTime_diff-spring-type2-type3.phpt]  XFAIL REASON: Various bugs exist
> DateTime::diff() -- spring type3 type2 [ext/date/tests/DateTime_diff-spring-type3-type2.phpt]  XFAIL REASON: Various bugs exist
57a58,73
> SCCP 002: Conditional Constant Propagation of non-escaping array elements [ext/opcache/tests/opt/sccp_002.phpt]
> SCCP 003: Conditional Constant Propagation of non-escaping array elements [ext/opcache/tests/opt/sccp_003.phpt]
> SCCP 004: Conditional Constant Propagation of non-escaping array elements [ext/opcache/tests/opt/sccp_004.phpt]
> SCCP 005: Conditional Constant Propagation of non-escaping array elements [ext/opcache/tests/opt/sccp_005.phpt]
> SCCP 006: Conditional Constant Propagation of non-escaping array elements [ext/opcache/tests/opt/sccp_006.phpt]
> SCCP 007: Conditional Constant Propagation of non-escaping array elements [ext/opcache/tests/opt/sccp_007.phpt]
> SCCP 008: Conditional Constant Propagation of non-escaping array elements [ext/opcache/tests/opt/sccp_008.phpt]
> SCCP 009: Conditional Constant Propagation of non-escaping array elements [ext/opcache/tests/opt/sccp_009.phpt]
> SCCP 010: Conditional Constant Propagation of non-escaping object properties [ext/opcache/tests/opt/sccp_010.phpt]
> SCCP 011: Conditional Constant Propagation of non-escaping object properties [ext/opcache/tests/opt/sccp_011.phpt]
> SCCP 012: Conditional Constant Propagation of non-escaping object properties [ext/opcache/tests/opt/sccp_012.phpt]
> SCCP 031: Echo optimizations [ext/opcache/tests/opt/sccp_031.phpt]
> SCCP 032: Yield from optimizations [ext/opcache/tests/opt/sccp_032.phpt]
> SCCP 022: Invalid types [ext/opcache/tests/opt/sccp_022.phpt]                                           
> SCCP 024: Const replacing to op2 of INSTANCEOF [ext/opcache/tests/opt/sccp_024.phpt]
> Test match jump table optimizer [ext/opcache/tests/match/002.phpt]
61a78,81
> Test opcode breakpoints [sapi/phpdbg/tests/breakpoints_004.phpt]                                        
> Stepping with exceptions must not be stuck at CATCH [sapi/phpdbg/tests/stepping_001.phpt]
> Test breaks on HANDLE_EXCEPTION [sapi/phpdbg/tests/exceptions_003.phpt]                                 
> Test next command on function boundaries [sapi/phpdbg/tests/next_001.phpt]  
```

davidrans commented:
> ```diff
> SCCP 031: Echo optimizations [ext/opcache/tests/opt/sccp_031.phpt]
> ```

That could be related

djmetzle commented:
@davidrans have you been leveraging [our docker container](https://github.com/iFixit/server-templates/tree/master/rpm-repo/build-scripts/danny-php) to build and test this?

The version gets checked out here: https://github.com/iFixit/server-templates/blob/c4bdbc47dd709d738e044c6697a47dd7a52b1095/rpm-repo/build-scripts/danny-php/scripts/build-php.sh#L35

davidrans commented:
I haven't! I just did it on Cominor.

davidrans committed:
SCCP: optimize ZEND_ECHO_ESCAPE the same as ZEND_ECHO

The compiler tries to figure out at compile time which `echo`
statements are, for example, going to be noops. The same rules should
apply to ECHO_ESCAPE as well.

davidrans committed:
PHPDBG: fix opcode parsing bug

This was a pre-existing PHPDBG bug that our new opcode exposed. PHPDBG
would not parse opcodes correctly if they were more than two words
separated by underscores (as many opcodes, like ZEND_POST_INC, are).

davidrans commented:
```diff
▶  bash -c 'diff <(sort original) <(sort patched3)'
53c53
< Number of tests : 16100             10957
---
> Number of tests : 16103             10960
62c62
< Tests passed    : 10921 ( 67.8%) ( 99.7%)
---
> Tests passed    : 10924 ( 67.8%) ( 99.7%)
65c65
< Time taken      :   87 seconds
---
> Time taken      :   63 seconds
```

All the same tests are now passing on this branch as on the source!
QA :rocket:

danielbeardsley commented:
CR :+1: 

I think these two extra commits might be worth keeping as separate commits, but we may want to update the README to include cherry-picking these as well.

davidrans commented:
Just updated the README
QA :rocket:

davidrans commented:
Closing in favor of https://github.com/iFixit/php-src/pull/21. PHP's releases page is confusing and I didn't realize 8.0.3 was the most recent stable version.
